File: factory-td/src/Camera.cpp
// =====================================================================
// Camera.cpp —— 摄像机实现（移植自 Python core/Camera.py）
// 坐标变换使用运行时屏幕尺寸（窗口缩放/最大化后保持点击与世界坐标一致）
// =====================================================================
#include "Camera.h"
#include <algorithm>
#include <cmath>

void Camera::init() {
    // Python: 初始定位地图中心（使用当前屏幕尺寸）
    x = (cfg::GRID_WIDTH * cfg::TILE_SIZE) / 2.0f - screenW / 2.0f;
    y = (cfg::GRID_HEIGHT * cfg::TILE_SIZE) / 2.0f - screenH / 2.0f;
    targetX = x;
    targetY = y;
}

void Camera::update(float dt) {
    // 帧率无关的指数平滑：把"每帧固定插值 0.1"换算成"每秒速率 CAMERA_SMOOTH_RATE"。
    //   每帧系数 k = 1 - exp(-rate * dt)
    // 60 帧时 k≈0.095（与旧手感一致）；144 帧时自动变小，总位移不随帧率变化。
    const float k = 1.0f - std::exp(-cfg::CAMERA_SMOOTH_RATE * dt);
    x += (targetX - x) * k;
    y += (targetY - y) * k;
    // 钳制到地图左上角（Python只钳制>=0）
    x = std::max(0.0f, x);
    y = std::max(0.0f, y);
}

void Camera::move(float dx, float dy, float dt) {
    // Python: 移动速度除以缩放，缩小时世界移动更慢
    // speedScale：启动菜单中的摄像机速度档位（慢0.6 / 标准1.0 / 快1.6）
    // 速度是"像素/秒"，乘 dt 后与帧率无关（旧写法按帧跑，144Hz 下会快 2.4 倍）
    const float speed = cfg::CAMERA_SPEED_PER_SEC * speedScale * dt;
    targetX += dx * speed / zoom;
    targetY += dy * speed / zoom;
}

sf::Vector2f Camera::worldToScreen(float wx, float wy) const {
    return {wx * zoom + (screenW / 2.0f - x * zoom),
            wy * zoom + (screenH / 2.0f - y * zoom)};
}

sf::Vector2f Camera::screenToWorld(float sx, float sy) const {
    return {(sx - (screenW / 2.0f - x * zoom)) / zoom,
            (sy - (screenH / 2.0f - y * zoom)) / zoom};
}

File: python版（老版）/core/Camera.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# Author : JYGAME
# Copyright : JYGAME

"""摄像机类

负责管理游戏视角的移动和缩放
"""

import pygame
from settings import CAMERA_SPEED, ZOOM_MIN, ZOOM_MAX, ZOOM_SPEED, SCREEN_WIDTH, SCREEN_HEIGHT, MAP_WIDTH, MAP_HEIGHT, TILE_SIZE


class Camera:
    """摄像机类"""
    
    def __init__(self):
        """初始化摄像机"""
        # 摄像机位置（世界坐标）- 初始化为地图中心
        self.x = (MAP_WIDTH * TILE_SIZE) // 2 - SCREEN_WIDTH // 2
        self.y = (MAP_HEIGHT * TILE_SIZE) // 2 - SCREEN_HEIGHT // 2
        
        # 缩放比例
        self.zoom = 1.0
        
        # 目标位置（用于平滑移动）- 初始化为地图中心
        self.target_x = self.x
        self.target_y = self.y
        
        # 移动速度
        self.speed = CAMERA_SPEED
    
    def update(self, delta_time: float) -> None:
        """更新摄像机位置
        
        Args:
            delta_time: 帧时间间隔
        """
        # 平滑移动到目标位置
        self.x += (self.target_x - self.x) * 0.1
        self.y += (self.target_y - self.y) * 0.1
        
        # 确保摄像机不超出地图边界
        self.x = max(0, self.x)
        self.y = max(0, self.y)
    
    def move(self, dx: float, dy: float) -> None:
        """移动摄像机
        
        Args:
            dx: x方向移动量
            dy: y方向移动量
        """
        self.target_x += dx * self.speed / self.zoom
        self.target_y += dy * self.speed / self.zoom
    
    def zoom_in(self) -> None:
        """放大摄像机"""
        self.zoom = min(self.zoom + ZOOM_SPEED, ZOOM_MAX)
    
    def zoom_out(self) -> None:
        """缩小摄像机"""
        self.zoom = max(self.zoom - ZOOM_SPEED, ZOOM_MIN)
    
    def get_offset(self) -> tuple:
        """获取屏幕偏移量
        
        Returns:
            屏幕中心相对于摄像机位置的偏移量
        """
        return (
            SCREEN_WIDTH // 2 - self.x * self.zoom,
            SCREEN_HEIGHT // 2 - self.y * self.zoom
        )
    
    def world_to_screen(self, world_x: float, world_y: float) -> tuple:
        """将世界坐标转换为屏幕坐标
        
        Args:
            world_x: 世界x坐标
            world_y: 世界y坐标
        
        Returns:
            屏幕坐标(x, y)
        """
        offset_x, offset_y = self.get_offset()
        return (
            world_x * self.zoom + offset_x,
            world_y * self.zoom + offset_y
        )
    
    def screen_to_world(self, screen_x: float, screen_y: float) -> tuple:
        """将屏幕坐标转换为世界坐标
        
        Args:
            screen_x: 屏幕x坐标
            screen_y: 屏幕y坐标
        
        Returns:
            世界坐标(x, y)
        """
        offset_x, offset_y = self.get_offset()
        return (
            (screen_x - offset_x) / self.zoom,
            (screen_y - offset_y) / self.zoom
        )
    
    def apply(self, rect: pygame.Rect) -> pygame.Rect:
        """将矩形从世界坐标转换到屏幕坐标
        
        Args:
            rect: 世界坐标中的矩形
        
        Returns:
            屏幕坐标中的矩形
        """
        x, y = self.world_to_screen(rect.x, rect.y)
        return pygame.Rect(
            x, y,
            rect.width * self.zoom,
            rect.height * self.zoom
        )